File: problems/GetAllElements.cpp
/**
 * Definition for a binary tree node.
 */
#include<vector>
#include<stack>
using namespace std;

struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

class Solution {
public:
    vector<int> getAllElements(TreeNode* root1, TreeNode* root2) {

        vector<int> vec;
        addAll(root1, vec);
        addAll(root2, vec);
        sort(vec.begin(), vec.end());
        return vec;
    }

    void addAll(TreeNode* r, vector<int>& into){
        if (r == nullptr ) {
            return;
        }

        stack<TreeNode*> s;

        s.push(r);

        while(!s.empty()){
            TreeNode* node = s.top();
            s.pop();

            into.push_back(node->val);

            if(node->left) {
                s.push(node->left);
            }

            if(node->right) {
                s.push(node->right);
            }
        }
    }
};

File: sept_challenge/src/solution/get_all_elements.rs
struct Solution {}
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn get_all_elements(
        root1: Option<Rc<RefCell<TreeNode>>>,
        root2: Option<Rc<RefCell<TreeNode>>>,
    ) -> Vec<i32> {
        let mut ans: Vec<i32> = Vec::new();
        if let Some(r) = root1 {
            Solution::traverse(Some(&r), &mut ans);
        }

        if let Some(r) = root2 {
            Solution::traverse(Some(&r), &mut ans);
        }

        ans.sort();
        ans
    }

    fn traverse(root: Option<&Rc<RefCell<TreeNode>>>, vec: &mut Vec<i32>) -> () {
        if let Some(node) = root {
            Solution::traverse(node.borrow().left.as_ref(), vec);
            vec.push(node.borrow().val);
            Solution::traverse(node.borrow().right.as_ref(), vec);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ex1() {
        assert_eq!(2 + 2, 4);
    }
}
